kimchi/src/prover.rs

```diff
@@ -424,79 +395,175 @@ where
+            }
+
+            drop(emul8);
+
+            // endoscaling scalar computation
+            let emulscalar8 =
+                endomul_scalar::constraint(range::ENDOMUL_SCALAR.start).evaluations(&env);
+            t8 += &emulscalar8;
+
+            if cfg!(test) {
+                let (_, res) = emulscalar8
+                    .clone()
+                    .interpolate()
+                    .divide_by_vanishing_poly(index.cs.domain.d1)
+                    .unwrap();
+                assert!(res.is_zero());
+            }
+
+            drop(emulscalar8);
+
+            // poseidon
+            let pos8 = poseidon::constraint().evaluations(&env);
+            t8 += &pos8;
+
+            if cfg!(test) {
+                let (_, res) = pos8
+                    .clone()
+                    .interpolate()
+                    .divide_by_vanishing_poly(index.cs.domain.d1)
+                    .unwrap();
+                assert!(res.is_zero());
+            }
+
+            drop(pos8);
+
+            // chacha
+            if index.cs.chacha8.as_ref().is_some() {
                 let chacha = chacha::constraint(range::CHACHA.start).evaluations(&env);
                 t4 += &chacha;
-                drop(chacha);
-                t4
+
+                if cfg!(test) {
+                    let (_, res) = chacha
+                        .clone()
+                        .interpolate()
+                        .divide_by_vanishing_poly(index.cs.domain.d1)
+                        .unwrap();
+                    assert!(res.is_zero());
+                }
             }
-        };
 
-        // quotient polynomial for lookup
-        let (t4, t8) = match index.cs.lookup_constraint_system.as_ref() {
-            None => (t4, t8),
-            Some(lcs) => combine_evaluations(
-                (t4, t8),
-                alpha,
-                alphas[alphas.len() - 1],
-                lookup::constraints(&lcs.dummy_lookup_values[0], index.cs.domain.d1)
-                    .iter()
-                    .map(|e| e.evaluations(&env))
-                    .collect(),
-            ),
-        };
+            // lookup
+            if let Some(lcs) = index.cs.lookup_constraint_system.as_ref() {
+                let mut alpha_pow = alphas[alphas.len() - 1];
+                let pows = (0..).map(|_| {
+                    alpha_pow *= alpha;
+                    alpha_pow
+                });
 
-        // divide contributions with vanishing polynomial
-        let (mut t, res) = (&(&t4.interpolate() + &t8.interpolate()) + &public_poly)
-            .divide_by_vanishing_poly(index.cs.domain.d1)
-            .map_or(Err(ProofError::PolyDivision), Ok)?;
-        if !res.is_zero() {
-            return Err(ProofError::PolyDivision);
-        }
+                let constraints =
+                    lookup::constraints(&lcs.dummy_lookup_values[0], index.cs.domain.d1);
+                for (constraint, alpha_pow) in constraints.into_iter().zip_eq(pows) {
```

Comment:
I'm confused by `zip_eq` here; isn't the alpha iterator infinite?

Reply:
good catch, this is an artifact of the alpha iterator PR (since I spill out the correct number of powers of alpha, we can use a `zip_eq`). But I'm confused how the tests are passing with this.